libtracepoint-control-cpp/include/tracepoint/TracingSession.h

```diff
@@ -0,0 +1,709 @@
+
+        Flags use the perf_event_sample_format values defined in <linux/perf_event.h>
+        or <tracepoint/PerfEventAbi.h>.
+        
+        The default value is:
+
+        | PERF_SAMPLE_TID
+        | PERF_SAMPLE_TIME
+        | PERF_SAMPLE_CPU
+        | PERF_SAMPLE_RAW
+
+        The following flags are supported:
+
+        | PERF_SAMPLE_IDENTIFIER
+        | PERF_SAMPLE_IP
+        | PERF_SAMPLE_TID
+        | PERF_SAMPLE_TIME
+        | PERF_SAMPLE_ADDR
+        | PERF_SAMPLE_ID
+        | PERF_SAMPLE_STREAM_ID
+        | PERF_SAMPLE_CPU
+        | PERF_SAMPLE_PERIOD
+        | PERF_SAMPLE_CALLCHAIN
+        | PERF_SAMPLE_RAW
+
+        Note that you'll almost always want to include PERF_SAMPLE_RAW since that
+        is the event's raw data (the event field values).
+        */
+        constexpr TracingSessionOptions&
+        SampleType(uint32_t sampleType) noexcept
+        {
+            m_sampleType = sampleType & SampleTypeSupported;
+            return *this;
+        }
+
+        /*
+        For realtime sessions only: sets the number of bytes of unconsumed event
+        data (counting both SAMPLE and non-SAMPLE events) that a realtime buffer
+        must contain to trigger wakeup (see WaitForWakeup).
+
+        The default value is WakeupWatermark(0).
+
+        Note that wakeup conditions are evaluated per-buffer. For example, if 3
+        buffers each contain 32760 bytes of pending data, none of them would
+        trigger a WakeupWatermark(32768) condition.
+        */
+        constexpr TracingSessionOptions&
+        WakeupWatermark(uint32_t wakeupWatermark) noexcept
+        {
+            m_wakeupUseWatermark = true;
+            m_wakeupValue = wakeupWatermark;
+            return *this;
+        }
+
+    private:
+
+        friend class TracingSession;
+
+    private:
+
+        size_t const m_bufferSize;
+        TracingMode const m_mode;
+        bool m_wakeupUseWatermark;
+        uint32_t m_wakeupValue;
+        uint32_t m_sampleType;
+    };
+
+    /*
+    Manages a tracepoint collection session.
+
+    Basic usage:
+
+        TracingCache cache;
+        TracingSession session(
+            cache, // A metadata cache to use for this session.
+            TracingSessionOptions(TracingMode::RealTime, 65536) // Required settings
+                .SampleType(PERF_SAMPLE_TIME | PERF_SAMPLE_RAW) // Optional setting
+                .WakeupWatermark(32768)                         // Optional setting
+                );
+
+        error = session.EnableTracepoint("user_events", "MyFavoriteTracepoint");
+        // ... check for error.
+
+        error = session.EnableTracepoint("user_events", "MySecondTracepoint");
+        // ... check for error.
+
+        for (;;)
+        {
+            // Wait until one or more of the buffers reaches 32768 bytes of event data.
+            error = session.WaitForWakeup();
+            // ... check for error. Don't get into a busy loop if waiting fails!
```

Comment:
If this is important, you should probably add it to the example rather than a comment.

Reply:
I don't have a clear idea of how to add it without doing error handling. The idea is to indicate that error handling here is very important and that the error can't be ignored because it might get into a busy loop.